### PR TITLE
=htc #17403 fix connection pool reusing existing connections too early

### LINKED ISSUE #17403: HTTP client: If a client application does not read a response body, then future requests will stall
(Closely related to #16898 -- this is basically the client-side mirror of that issue.)

I have some HTTP client code which makes repeated requests to the same host.
After upgrading this code to use the pooled-connection HTTP client in 1.0-RC1, I have found that some requests never complete, even though the remote host has replied.

I believe that the issue is that my client code does not read the full response body in all cases. For example, where it can determine that the response is an error from the status code or the Content-Type it does not read the full response body.

I assume that the HTTP client pool does not notice that the TCP connection "down" stream is stalled in this way, and puts it back in the pool for reuse. The next request to the same host is sent on the "up" stream of that connection, but the next response is never read, as the "down" stream is stalled.

The following test demonstrates the issue:

```
import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.{HttpResponse, StatusCodes, HttpRequest}
import akka.stream.ActorFlowMaterializer
import org.specs2.mutable.Specification
import org.specs2.time.NoTimeConversions
import scala.concurrent.{Future, Await}
import scala.concurrent.duration._

class StalledConnectionSpec
  extends Specification
  with NoTimeConversions {

  implicit val system = ActorSystem()
  implicit val fm = ActorFlowMaterializer()

  "HTTP clients who do not read the response body" should {

    "not stall future requests to the same host" in {

      val request = HttpRequest(uri = "http://en.wikipedia.org/wiki/Main_Page")
      val httpClient: HttpRequest => Future[HttpResponse] = Http().singleRequest(_)

      // Make a request, don't read the body:
      val response1 = Await.result(httpClient(request), 10.seconds)
      response1.status mustEqual StatusCodes.OK

      // Make a second request to the same host:
      // The same TCP connection will be re-used by the pool, but it
      // is now blocked waiting for the body of `response1` to be read.
      // The request will time out waiting for that...
      val response2 = Await.result(httpClient(request), 10.seconds)
      response2.status mustEqual StatusCodes.OK
    }
  }
}
```

I think that the TCP connection should not be released to the pool for reuse until the full response has been read by the client code.

/cc @sirthias

### PR DESCRIPTION
**Pull request by sirthias**
Fixes #17403

/cc @jrudolph

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala [hunk 1]
```diff
@@ -67,10 +72,15 @@ private object PoolConductor {
       val doubler = Flow[SwitchCommand].mapConcat(x ⇒ x :: x :: Nil) // work-around for https://github.com/akka/akka/issues/17004
       val route = b.add(new Route(slotCount))
       val retrySplit = b.add(new RetrySplit())
+      val flatten = Flow[RawSlotEvent].mapAsyncUnordered(slotCount) {
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala]: Is the use of unordered correct here?

Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala]: or there can be at most one Event in circulation per slot?

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala [hunk 2]
```diff
@@ -22,15 +24,16 @@ private object PoolSlot {
...
   final case class ResponseDelivery(response: ResponseContext) extends ProcessorOut
-  sealed trait SlotEvent extends ProcessorOut
-  sealed trait SimpleSlotEvent extends SlotEvent
+  sealed trait RawSlotEvent extends ProcessorOut
+  sealed trait SlotEvent extends RawSlotEvent
   object SlotEvent {
-    final case class RequestCompleted(slotIx: Int) extends SimpleSlotEvent
-    final case class Disconnected(slotIx: Int, failedRequests: Int) extends SimpleSlotEvent
-    final case class RetryRequest(rc: RequestContext) extends SlotEvent
+    final case class WhenRequestCompleted(future: Future[RequestCompleted]) extends RawSlotEvent
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala]: the name is a bit misleading. You can just call it RequestCompletedFuture

**Comment by drewhk** [comment]
LGTM, if mapAsyncUnordered is correct here.

**Comment by drewhk** [comment]
@sirthias can I merge this? Is mapAsyncUnordered OK?

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala [hunk 3]
```diff
@@ -149,8 +152,21 @@ private object PoolSlot {
...
+          case x: HttpEntity.Strict ⇒ x -> FastFuture.successful(())
+          case x: HttpEntity.Default ⇒
+            val (newData, whenCompleted) = StreamUtils.captureTermination(x.data)
+            x.copy(data = newData) -> whenCompleted
+          case x: HttpEntity.CloseDelimited ⇒
+            val (newData, whenCompleted) = StreamUtils.captureTermination(x.data)
+            x.copy(data = newData) -> whenCompleted
+          case x: HttpEntity.Chunked ⇒
+            val (newChunks, whenCompleted) = StreamUtils.captureTermination(x.chunks)
+            x.copy(chunks = newChunks) -> whenCompleted
```
Comment (rkuhn) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala]: This does not handle all entity types: what about IndefiniteLength?

Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala]: `IndefiniteLength` is not a `ResponseEntity`.

Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala]: `HttpEntity.IndefiniteLength` is not a `ResponseEntity`. Therefore we won't see it here.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala (hunk 1 above)
Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala]: Unordered is fine. We don't want a slow response on one connection block us from seeing an update from another connection.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala (hunk 2 above)
Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala]: Ok, will do.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala (hunk 3 above)
Comment (rkuhn) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala]: I am massively confused by the type hierarchy here, it seems, but I trust you.

**Comment by rkuhn** [comment]
LGTM (modulo the point Endre raised about the mapAsyncUnordered—I have a hunch that it is fine but there may be dirty details there)

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala (hunk 3 above)
Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala]: It's all really simple... :)
4 super types (`RequestEntity`, `ResponseEntity`, `BodyPartEntity` and `UniversalEntity`) "select" sub sets of the 4 actual implementations: `Strict`, `Default`, `CloseDelimited` and `IndefiniteLength`.
Maybe we should provide a handy matrix in the docs that shows with super types comprises which implementation.

Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala]: Actually its five implementations, `Chunked` was missing from my list above.
We definitely need this matrix.... :)

**Comment by sirthias** [comment]
Ok, I renamed `WhenRequestCompleted` as suggested by @drewhk and rebased onto the current release-2.3-dev.
AFAICS there really is no issue with the `mapAsyncUnordered`. The responses from a single connection are naturally ordered and there is no connection between connections (as funny as that may sound). And even if a future for one connection should somehow overtake a previous future for the same connection then this wouldn't be a problem.

So, relax and smell the roses. :)

**Comment by drewhk** [comment]
Wonderful :) Will merge after validation.

File: akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala (hunk 3 above)
Comment (rkuhn) [inline akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala]: My problem when reading the code on github was that it is not at all obvious how Chunked extends any of RequestEntity or ResponseEntity, but I might just have missed something.

**Comment by rkuhn** [comment]
Good point: the roses do indeed smell wonderful out in the garden :-)

**Pull request merged**